WOR-703 Part 1: changes to existing GCP billing project wizard

blakery opened the pull request:
[WOR-703](https://broadworkbench.atlassian.net/browse/WOR-703)

This PR is the changes the existing step wizard for GCP, in order to extract it into a separate component.
Original PR [here](https://github.com/DataBiosphere/terra-ui/pull/3767)

<!---
Hello, friend!
Remember to mention what sort of testing/verification you performed in the course of building this PR, i.e. checking functionality in the browser locally.
Also, so if a screen recording [1] and/or screenshots would be a helpful way to communicate context, please consider including some in your PR description.
Thanks!

[1] https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac
-->


[WOR-703]: https://broadworkbench.atlassian.net/browse/WOR-703?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

blakery committed:
cherry-picking changes to existing GCP billing project wizard

blakery committed:
fixed alignment and restored state persistence

blakery committed:
fixed tests

cahrens commented on src/pages/billing/models/AzureManagedAppCoordinates.ts:
```diff
@@ -0,0 +1,7 @@
+  tenantId: string // UUID as string
+  subscriptionId: string // UUID as string
+  managedResourceGroupId: string
+  region?: string
```
Do we ever have a region?

blakery replied:
It's in the BPM's api docs. I don't know if we use it or not.
Same situation for `applicationDeploymentName`. 

We could remove them, with the idea that this is only the fields returned by both BPM _and_ rawls, then extend it later, if necessary.

cahrens commented on src/pages/billing/models/AzureManagedAppCoordinates.ts:
```diff
@@ -0,0 +1,7 @@
+  subscriptionId: string // UUID as string
+  managedResourceGroupId: string
+  region?: string
+  applicationDeploymentName?: string
```
Do we ever have an applicationDeploymentName?

blakery replied:
This is in `src/pages/billing/CreateAzureBillingProjectModal.js`:
```
h(Select, {
            id,
            isMulti: false,
            placeholder: 'Select a managed application',
            isDisabled: errorFetchingManagedApps !== false || !!subscriptionIdError,
            value: chosenManagedApp,
            onChange: ({ value }) => setChosenManagedApp(value),
            options: _.map(application => {
              return {
                value: application,
                label: !!application.region ?
                  `${application.applicationDeploymentName} (${application.region})` :
                  application.applicationDeploymentName
              }
            }, managedApps)
          })
```

blakery committed:
removed unused fields from model

cahrens commented:
@blakery I'm seeing the first text for _Step 2_ appear to the right of _Step 2_ instead of below it.

![image](https://user-images.githubusercontent.com/484484/217847884-5cc84bc3-e74b-486a-a64b-9ba4d1094edf.png)

cahrens commented:
It looks like we lost the extra space at the bottom of Step 4, which I believe was there to allow the display of error messages. Here it is on prod:

<img width="676" alt="image" src="https://user-images.githubusercontent.com/484484/217848350-f4733075-a020-4979-a747-27eae6444c99.png">

cahrens commented:
Test for _Step 4_ isn't appearing below _Step 4_, and the button text has changed to all caps (was that intentional?):

![image](https://user-images.githubusercontent.com/484484/217848677-585316b9-6a59-4a3a-bb19-df762fff41fd.png)


On prod:

<img width="1351" alt="image" src="https://user-images.githubusercontent.com/484484/217848751-479800d6-526a-44b4-8ae1-483055c21693.png">

cahrens commented:
There is something odd going on with the rendering of the checkbox.

![image](https://user-images.githubusercontent.com/484484/217849177-de22573b-9977-459b-8454-6b6723a4d0c8.png)

cahrens commented:
Since we have the room, can we make the select fields wider so that they don't resize for the very commonly shown message about "Billing Project name is too short"?

![image](https://user-images.githubusercontent.com/484484/217851063-b9b556c6-02dc-41ee-ae00-91c4682c771f.png)

cahrens commented on src/pages/billing/CreateGCPBillingProject.ts:
```diff
@@ -8,13 +8,24 @@ import Events from 'src/libs/events'
+  setChosenBillingAccount: (BillingAccount) => void
+  billingProjectName?: string
+  setBillingProjectName: (string) => void
+  existing: any
```
I think `existing` is a list of strings.

blakery replied:
it is: I just hadn't revisited this file after the point where I figured that out

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.ts:
```diff
@@ -0,0 +1,119 @@
+            setRefreshed(true)
+          } else {
+            setActiveStep(3)
+            //await authorizeAndLoadAccounts()
```
Why is line 67 here? Did it used to call that?

blakery replied:
It's leftover from when I was teasing apart logic.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.test.ts:
```diff
@@ -30,7 +39,7 @@ const getStep3AddTerraAsUserText = () => textMatcher('Add terra-billing@terra.bi
   "Billing Account User to your organization's billing account.")
 
-const getStep4CreateButton = () => screen.queryByText('Create Terra Billing Project')
+const getStep4CreateButton = () => screen.queryByRole('button', { name: 'create-billing-project' })
```
Having this aria-label "create-billing-project" is not friendly-- it means the screen reader reads that instead of the button text. There shouldn't be an aria-label for something that doesn't need it.

What was wrong with getting the control by the visible text?

blakery replied:
The testing library documentation encourages retrieving elements  by role when possible. 
What I failed to realize at the time, was that `name` also picks up on the element text (or maybe there was some other reason it wasn't working at the time).

Removing the aria-label, in any case.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.test.ts:
```diff
@@ -44,12 +53,10 @@ const verifyChecked = item => expect(item).toBeChecked()
 
 const testStepActive = stepNumber => {
-  screen.getAllByRole('listitem').forEach((step, index) => {
+  screen.queryAllByTestId('Step').forEach((step, index) => {
```
What was wrong with getting it by role?

cahrens replied:
Ah, I think you changed it so these are no longer presented as being in an ordered list, and instead use region. An ordered list might make more sense here, since you are supposed to do them sequentially.

blakery replied:
Mostly, they stopped being `li` elements.
Also, this would get _all_ `li`s on the page, and I wanted to be able to use list elements in other components, if it made sense.

cahrens replied:
Yes, using a list (I would say `ol` instead of `ul`, but the `ul` in prod actually works well) makes much more sense when using a screen reader-- it tells you the number of steps, which step you are in, and the use of `aria-current` within it makes more sense.

cahrens replied:
@blakery within this page this wizard there is expected to only be a single list, so I don't see that being a problem.

I really think it's important to go back to using list as the semantic structure.

blakery replied:
I put it back to a list, after we established it's completely fine to nest lists in html.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.test.ts:
```diff
@@ -70,16 +77,21 @@ const testStep2HaveBillingChecked = () => {
   verifyUnchecked(getStep3BillingAccountNoAccessButton())
+  verifyDisabled(getStep3AddedTerraBillingButton())
   verifyUnchecked(getStep3AddedTerraBillingButton())
-  expect(getStep3AddTerraAsUserText()).not.toBeNull()
```
Why remove these asserts?

blakery replied:
TLDR; Going through my git history, it looks like it was mostly a mistake. 

Long form:
The intended result was to split up testStep3InitialState into two separate functions: one that was testing which version of step 3 was rendered (testAddTerraUserRenderedForStep3), and the other that was testing the state of the inputs (testAddTerraUserStep3UninitializedState). 
I think by the time I got to that point, I had been staring at the code for too long, b/c I missed a step in my process.
Initially I had 4 functions: the two here, one that the other version of step 3 was rendered, and one checking the inputs on the other version. But it didn't make sense to use the other two functions, since the checks were already rolled into different places.

- `expect(getStep3VerifyUserAdded()).toBeNull()` was moved to `testAddTerraUserRenderedForStep3` (line 82, now)
- `expect(getStep3ContactBillingAdministrator()).toBeNull()` was one of the things I flipped the logic of, and put into the function testing that the other version of the component was rendered
- `expect(getStep3AddTerraAsUserText()).not.toBeNull()` .... just got missed, I think, and I didn't notice because I was still getting the signal that the component overall was being rendered 

I'll put the missing two back in the function that's testing that the component overall is rendered.


TY for catching that, and please excuse my stream of consciousness above.

cahrens commented on src/pages/billing/models/BillingProject.ts:
```diff
@@ -0,0 +1,26 @@
+import { AzureManagedAppCoordinates } from 'src/pages/billing/models/AzureManagedAppCoordinates'
+
+
+export type CloudPlatform = 'GCP' | 'Azure'
```
In the server response, it's actually 'AZURE'.

```
{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-azure-dev-20221202091509","subscriptionId":"3efc5bdf-be0e-44e7-b1d7-c08931e3c16c","tenantId":"0cb7a640-45a2-4ed6-be9f-63519f86e04b"},"projectName":"azure-dev-2022-12-02","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-dev-jan23-20230123125907","subscriptionId":"3efc5bdf-be0e-44e7-b1d7-c08931e3c16c","tenantId":"0cb7a640-45a2-4ed6-be9f-63519f86e04b"},"projectName":"azure-dev-2023-01-23","roles":["Owner"],"status":"Ready"},{"billingAccount":"billingAccounts/01A82E-CA8A14-367457","cloudPlatform":"GCP","invalidBillingAccount":false,"projectName":"callisto-dev","roles":["User","Owner"],"status":"Ready"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: b23e9f9a-a39e-472d-be48-46a4d1c337e5","projectName":"CARAzureRawls","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: acdb4ac1-9bd3-4078-a686-6380f2b011b0","projectName":"CARAzureTake3","roles":["Owner"],"status":"Error"},{"billingAccount":"billingAccounts/00708C-45D19D-27AAFA","cloudPlatform":"GCP","invalidBillingAccount":false,"projectName":"CARBilling-2","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"GCP","invalidBillingAccount":false,"projectName":"CARBilling3","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: bf45a290-4da9-4dde-a939-1e5e414c4725","projectName":"CARBilling4","roles":["Owner"],"status":"Error"},{"billingAccount":"billingAccounts/00708C-45D19D-27AAFA","cloudPlatform":"GCP","invalidBillingAccount":true,"projectName":"CARBillingTest","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: c2ab9efd-8204-403d-84b1-4f5edd906785","projectName":"CARDebugLZ","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: 9504473f-561a-46e5-8093-d209a7f1bdc0","projectName":"CARDebugLZ2","roles":["Owner"],"status":"Error"},{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-dev-previ-20221213131637","subscriptionId":"df547342-9cfd-44ef-a6dd-df0ede32f1e3","tenantId":"fad90753-2022-4456-9b0a-c7e5b934e408"},"projectName":"CARDec13Work","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: 6b01b991-5a7a-4328-8961-1654057beddb","projectName":"CARDec22Azure","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: de966e9a-59d4-47ed-9610-8df7e59b16b9","projectName":"CARDec22Take2","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: 515abd5d-8887-4c7e-88ab-5419d534cf4a","projectName":"CARDec7WS","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: bb030783-7dd7-4012-8960-61386191026f","projectName":"CARDec9WOR","roles":["Owner"],"status":"Error"},{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-dev-previ-20230119110658","subscriptionId":"df547342-9cfd-44ef-a6dd-df0ede32f1e3","tenantId":"fad90753-2022-4456-9b0a-c7e5b934e408"},"projectName":"CARJan19bemis","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-dev-previ-20230131123105","subscriptionId":"df547342-9cfd-44ef-a6dd-df0ede32f1e3","tenantId":"fad90753-2022-4456-9b0a-c7e5b934e408"},"projectName":"CARJan31AzureWest","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-dev-previ-20221108090449","subscriptionId":"62b22893-6bc1-46d9-8a90-806bb3cce3c9","tenantId":"fad90753-2022-4456-9b0a-c7e5b934e408"},"projectName":"CARTerraNov8_ws1","roles":["Owner"],"status":"Ready"},{"billingAccount":"billingAccounts/01A82E-CA8A14-367457","cloudPlatform":"GCP","invalidBillingAccount":false,"projectName":"galaxy-anvil-dev","roles":["Owner"],"status":"Ready"},{"billingAccount":"billingAccounts/00708C-45D19D-27AAFA","cloudPlatform":"GCP","invalidBillingAccount":false,"projectName":"general-dev-billing-account","roles":["User"],"status":"Ready"},{"cloudPlatform":"AZURE","invalidBillingAccount":false,"managedAppCoordinates":{"managedResourceGroupId":"mrg-terra-dev-previ-20220930130036","subscriptionId":"3efc5bdf-be0e-44e7-b1d7-c08931e3c16c","tenantId":"0cb7a640-45a2-4ed6-be9f-63519f86e04b"},"projectName":"terra-dev-test-project","roles":["Owner"],"status":"Ready"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: 90282f4f-af84-479b-9472-02dc8533d510","projectName":"WorkspacesDemo","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: e140fc23-206f-428c-9978-468f854476b8","projectName":"WorkspacesDemo2","roles":["Owner"],"status":"Error"},{"cloudPlatform":"UNKNOWN","invalidBillingAccount":false,"message":"Unable to find billing profile in Billing Profile Manager for billing profile id: 651011ae-8006-45e2-91a3-fcc5adf66551","projectName":"WorkspacesDemo3","roles":["Owner"],"status":"Error"}]

cahrens replied:
'UNKNOWN' is also a possible value, when we can't retrieve the record from BPM. It's an error case, but Rawls does document it and I have a billing project in that state.

blakery replied:
Good point: I didn't really think about the unknown case.

cahrens commented on src/pages/billing/models/BillingProject.ts:
```diff
@@ -0,0 +1,26 @@
+  cloudPlatform: CloudPlatform
+  projectName: string
+  invalidBillingAccount: boolean
+  roles: string[]
```
Type for `roles` should be `BillingRole`.

I would also suggest just putting `BillingRole` directly in this file instead of having a separate `BillingRole.ts` file.

cahrens commented on src/pages/billing/models/BillingProject.ts:
```diff
@@ -0,0 +1,26 @@
+export interface GCPBillingProject extends BillingProject {
+  cloudPlatform: 'GCP'
+  billingAccount: string
+  servicePerimeter: string
```
I don't think `servicePerimeter` is returned for all GCP billing projects.

cahrens commented on src/pages/billing/models/BillingAccount.ts:
```diff
@@ -0,0 +1,5 @@
+export interface BillingAccount {
```
Is it necessary/preferred in typescript to have each of these definitions in their own file?

cahrens replied:
For clarity, can we call this `GoogleBillingAccount`? It took me a while to figure out what it was.

msilva-broad replied:
to your earlier question, current guidance is to have the type defined within the module that "owns" or emits it... In many cases, this should push types to be near the ajax "data-client" module, or a provider if one exists... But for cases where a number of modules make use of the type and there is no clear "owner", having a shared type be placed further up the directory tree is a good way to go.

cahrens replied:
@msilva-broad sorry, I'm not understanding how to apply your answer to this case. Do you think we should have all these small individual files for each definition?

cahrens replied:
Sorry, but I have to resolve this in order to be able to merge.

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/StepWizard.ts:
```diff
@@ -0,0 +1,18 @@
+export interface StepWizardProps {
+  title: string
+  intro: string
+  children?: React.ReactNode
```
I don't think we ever what the StepHeader to have children. The text is always supposed to be below it.

blakery committed:
update models

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/StepFields.ts:
```diff
@@ -0,0 +1,58 @@
+
+export const LabeledField = ({ label, formId, required = false, children = [], style }: LabeledFieldProps) => div({ style: { display: 'flex', flexDirection: 'column', ...style } }, [
+  h(FormLabel, { htmlFor: formId, required }, [label]),
+  children
```
Note that if you are able to put the children within the label, you don't need to link them together with an ID. So for example, the following is perfectly OK and avoids the need to create an ID:

````
<label>Address:<input></input></label>

blakery replied:
I think I had to do this because the project-wide `ValidatedInput` component uses some absolute positioning (and maybe some other witchery, too), so this was the only way I could get it to work.

The `CreateGCPBillingProject` component does essentially the same thing, I'm assuming for the same reason.

cahrens replied:
OK, that's fine. Just wanted to let you know because the general pattern in terra-ui is to always use an _id_, but that's not necessary if you can nest the components.

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/Step.ts:
```diff
@@ -0,0 +1,37 @@
+export interface StepProps {
+  isActive: boolean
+  title?: string
+  introText?: React.ReactNode[]
```
Why is `introText` a ReactNode (instead of a string)?

blakery replied:
So you could do things like text formatting, adding links, etc.

`ReactNode` is just anything react can render, which includes strings.

I switched to just composing it more explicitly using the `StepHeader` as the first child, which is more in-line with standard react patterns (I just checked, and all of the current usages do this).
I'll go ahead and remove title and introText as props, since they aren't really used anymore.

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/Step.ts:
```diff
@@ -0,0 +1,37 @@
+    width: '100%',
+    border: active ? `1px solid ${colors.accent()}` : `1px solid ${colors.accent(0.2)}`,
+    backgroundColor: active ? colors.light(0.5) : colors.light(0.3),
+    boxShadow: active ? '0 0 5px 0 rgba(77,114,170,0.5)' : 'none'
```
This shadow color was hard-coded before? Just wondering if it will look OK with other whitelabel themes.

blakery replied:
yes - I generally try *very* hard to avoid hardcoding values like that. I don't think I touched the colors after moving this from the original file.

I just plugged it into a converter. and the hex value came out to `#4d72aa80`
The 'accent' color value is `#4d72aa`, which translates to `rgba(77, 114, 170, 1)`
So should this be `colors.accent(.5)`?

blakery replied:
I just tested it out, and it looks correct / the same - so I'm going with it, barring objections.

cahrens replied:
Sounds good.

cahrens commented on src/pages/billing/NewBillingProjectWizard/StepWizard/Step.ts:
```diff
@@ -0,0 +1,37 @@
+  children?: React.ReactNode[]
+}
+
+export const Step = ({ isActive, title, ...props }: StepProps) => section({
```
As noted above, go back to using `li` instead of section (with parent `ol` or `ul`) so screen reader users are provided information about the number of steps.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/AddTerraAsBillingAccountUserStep.test.ts:
```diff
@@ -0,0 +1,138 @@
+    })
+
+    it('shows the correct instructions to the user', () => {
+      const instructions = textMatcher('Add terra-billing@terra.bio as a Billing Account User to your billing account.')
```
Testing the full text that is displayed feels like a bit too much to me (as well as testing that some controls are radio buttons). @msilva-broad do you think it is possible for unit tests to get _too_ specific/detailed?

blakery replied:
I mostly was trying to break out the step-specific logic from the main test file (largely to try and figure out what was actually going on). I think that selector was taken directly from there.
I'm not at all attached to it.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/BillingAccountAccessStep.ts:
```diff
@@ -0,0 +1,43 @@
+}
+
+export const BillingAccountAccessStep = ({ isActive, ...props }: BillingAccountAccessStepProps) => h(Step, { isActive }, [
+  h(StepHeader, { title: 'STEP 2', children: ['Select an existing billing account or create a new one.'] }),
```
It would be better if the legend could be `Select an existing billing account or create a new one.` vs the explanatory text below, as the radio buttons really go with `Select an existing billing account or create a new one.`.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GoToGCPConsoleStep.ts:
```diff
@@ -0,0 +1,39 @@
+        ...Utils.newTabLinkProps,
+        onClick: () => {
+          Ajax().Metrics.captureEvent(Events.billingGCPCreationStep1)
+          // FIXME: this seems wrong
```
I think this might be tied into to the fact that we didn't want to force people to click on the button, and we want people to be able to interact with the previous controls even when they are on a different step. It's all pretty complicated though.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GoToGCPConsoleStep.ts:
```diff
@@ -0,0 +1,39 @@
+  return h(Step, { isActive }, [
+    StepHeader({ title: 'STEP 1' }),
+    h(StepFields, [
+      h(StepFieldLegend, ['Go to the Google Cloud Platform Billing Console and sign-in with the same user you use to login to Terra.']),
```
I don't think this should be in a legend/field set, as there are no form options.

blakery replied:
good point
I think when converting to standardized components I just went a bit ham.

cahrens replied:
I changed this by introducing `StepFieldInfo`, which might not be the greatest way to do handle this, but easily allowed consistent styling without forcing everything to be in a field set.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.ts:
```diff
@@ -0,0 +1,119 @@
+        isFinished: activeStep > 3,
+      }) :
+      h(ContactAccountAdminToAddUserStep, {
+        //persistenceId,
```
Line 75?

blakery replied:
right: I tried to move some state to the component at one point, but didn't clean that up

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.ts:
```diff
@@ -0,0 +1,119 @@
+            await authorizeAndLoadAccounts()
+            setActiveStep(4)
+            // I think this makes sense here, since we just refreshed auth and accounts,
+            // but the existing unit tests fail with it, since the refresh button in step 4 doesn't come up
```
I'm not understanding-- the unit test expects the refresh button to come up at this point?

blakery replied:
yes.
when actually going through the process as a user, when you check that you've verified that the user was added, it refreshes the auth and billing accounts (note that the text displayed on the checkbox label includes `(requires reauthentication)` at the end - so this appears to be the intended progression).
Then if we still don't have billing accounts, the user has to refresh again, from step 4 before the error message that there are no billing accounts, even though they _just_ refreshed

the unit tests basically test that step 4 is in the default/uninitialized state after step 3

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/CreateTerraProjectStep.ts:
```diff
@@ -0,0 +1,143 @@
+      marginTop: '0.25rem'
+    }
+  }),
+  // TODO: can we get away without the refreshed prop, and just rely on whether or not we've gotten to this step?
```
The goal with "refreshed" was that we don't want to keep showing the error each time, and instead what to present them with the UI to contact support once they have attempted to refresh once.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/ContactAccountAdminToAddUserStep.ts:
```diff
@@ -0,0 +1,50 @@
+      span({ style: { fontWeight: 'bold' } }, [' to your organization\'s billing account.'])
+    ]),
+    div({ style: { width: '30%', display: 'flex', flexDirection: 'row' } }, [
+      h(LabeledCheckbox, {
```
There is something not correct about how the state persistence is working. On production, if I check this checkbox and then refresh the page, it remembers that I have checked the checkbox and that I am on Step 4. On your branch, after I refresh the checkbox is cleared.

blakery replied:
I had disabled the state storage so I could actually test behavior - when I put it back, the variable name had changed.

On a related note: should we clear all of this user state when they finish the last step? 
I found it really obnoxious that I'd come back and it would still have all the previous state from the last project.

Then again, the current implementation is really only intended to walk them through the first project. So maybe that's just something to revisit when we switch to using this instead of the modal.

cahrens replied:
Right, with the current implementation you only see the wizard if you don't have any billing projects.

marctalbott commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/AddTerraAsBillingAccountUserStep.ts:
```diff
@@ -0,0 +1,63 @@
+        'Add ',
+        span({ style: { fontWeight: 'bold' } }, ['terra-billing@terra.bio']),
+        ' as a Billing Account User',
+        span({ style: { fontWeight: 'bold' } }, [' to your billing account.']), //]),
```
nit: remove commented code

cahrens replied:
Done.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/CreateTerraProjectStep.ts:
```diff
@@ -0,0 +1,143 @@
+          h(ButtonPrimary, {
+            onClick: submit,
+            role: 'button',
+            'aria-label': 'create-billing-project',
```
As mentioned in the test, remove the `aria-label` from this.

cahrens commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/CreateTerraProjectStep.ts:
```diff
@@ -0,0 +1,143 @@
+        div({ style: { display: 'flex', flexDirection: 'column', alignItems: 'center' } }, [
+          h(ButtonPrimary, {
+            onClick: submit,
+            role: 'button',
```
I'm pretty sure you don't need to add the button role yourself-- through some dark magic that I can't locate now in code, it gets added for you (unless you have an href, in which case this gets rendered as an anchor tag).

blakery replied:
It wasn't getting added. I first detected it when I switched the selector in the test to use select by role. 

I think that's why I ended up adding the 'aria-label': I assumed that the button would be selectable as a button, and was ready to gouge out my eyeballs trying to figure out why it wouldn't pick it up.

I ended up examining the DOM. and saw that the component wasn't getting the role assigned. I specified it explicitly, and the selector picked it up.

cahrens replied:
That's weird because when I removed line 78 locally, it still had the role. And it has the role on production, even though it isn't specified explicitly in the old code. Can you try again?

<img width="703" alt="image" src="https://user-images.githubusercontent.com/484484/218129432-6f79baf6-1fd1-4654-a1df-2c50a5dd25e8.png">

blakery replied:
It's working now, for some reason.

blakery committed:
more model and type fixes

blakery committed:
remove unhelpful aria-label

blakery committed:
put back checks that were removed by mistake

blakery committed:
put verifiedUsersAdded back into user state being stored

blakery committed:
move instructive text to form legend

blakery committed:
remove hard-coded color value

blakery committed:
removed unused props from Step component

all usages now use StepHeader explicitly

blakery committed:
switch steps back to a list

turns out, a list with in a list is perfectly ok: so my concern about that restricting the usage of the components was unfounded

blakery committed:
move descriptive text back into legend to fix layout

blakery committed:
remove explicit role

it works without it, now

blakery committed:
give step 4 a little more space so it doesn't resize on errors, and improve button placement

blakery committed:
restore margin overrides to fix checkbox label and layout

cahrens commented:
I'm changing the layout so that the billing project selects are wider and thus the error messages below don't have to wrap. For example:

![image](https://user-images.githubusercontent.com/484484/218508019-58bf5401-f204-446c-9e89-612d3f350479.png)

cahrens committed:
Layout fixes.

cahrens committed:
Remove blank lines, ensure aria-checked is always included.

marctalbott commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GoToGCPConsoleStep.ts:
```diff
@@ -0,0 +1,40 @@
+        ['Go to the Google Cloud Platform Billing Console and sign-in with the same user you use to login to Terra.']),
+      h(ButtonOutline, {
+        disabled: false,
+        href: 'https://console.cloud.google.com',
```
I think https://console.cloud.google.com/billing would be a more useful link here. Changing the link may be out of scope, so it's up to you

cahrens replied:
Happy to change the link-- @admull do you agree with @marctalbott's proposal?

cahrens replied:
Decided with Adam to keep as-is.

marctalbott commented on src/pages/billing/NewBillingProjectWizard/GCPBillingProjectWizard/GCPBillingProjectWizard.ts:
```diff
@@ -0,0 +1,118 @@
+        setAccessToBillingAccount(access)
+        setAccessToAddBillingAccountUser(undefined)
+        setVerifiedUsersAdded(undefined) // reset verification of step 3
+        setRefreshed(false) // this might or might not be necessary
```
Is it necessary?

cahrens replied:
I don't know… the comment was from @blakery .

cahrens replied:
Sorry, I have to resolve this to be able to merge….

marctalbott reviewed (approved):
There's currently only tests for the `AddTerraAsBillingAccountUserStep`, do we need tests for each step? Tested it out locally though and everything looks good

cahrens commented:
> There's currently only tests for the `AddTerraAsBillingAccountUserStep`, do we need tests for each step? Tested it out locally though and everything looks good

The was an extensive pre-existing unit test for the whole wizard that is still running. I'm not sure what @blakery's plan was with regard to testing each individual step in isolation, but we certainly haven't lost any coverage with this PR.

cahrens committed:
Code review feedback.
